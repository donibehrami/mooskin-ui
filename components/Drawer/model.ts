import { IDivBoxComponentProps } from '../Box/model';

export type IContentSize = 'xs' | 'sm' | 'md' | 'lg' | 'xl' | 'full';

export interface IDrawerComponentProps extends IDivBoxComponentProps {
    /** whether the Drawer is open or not */
    isOpen?: boolean;

    /** placement of the Drawer */
    placement?: 'bottom' | 'left' | 'right' | 'top';

    /** size of the Drawer */
    size?: IContentSize;

    /** call close callback when esc key is pressed */
    closeOnEsc?: boolean;

    /** call close callback when the overlay is clicked */
    closeOnOverlayClick?: boolean;

    /** close callback */
    onClose?: (e: React.MouseEvent<HTMLDivElement>) => void;
}

export interface IDrawerContentComponentProps extends IDivBoxComponentProps {
    /** whether the Drawer is open or not */
    isOpen?: boolean;

    /** size of the Drawer */
    size?: IContentSize;

    /** placement of the Drawer Content */
    placement?: 'bottom' | 'left' | 'right' | 'top';
}

export interface IDrawerOverlayComponentProps extends IDivBoxComponentProps {
    /** whether the Drawer is open or not */
    isOpen?: boolean;

    /** callback when clicking the drawer overlay */
    onClickOverlay?: (e: React.MouseEvent<HTMLDivElement>) => void;
}

export interface IDrawerCloseButtonComponentProps extends IDivBoxComponentProps {
    /** callback when clicking the drawer overlay */
    onClickButton?: (e: React.MouseEvent<HTMLDivElement>) => void;
}
