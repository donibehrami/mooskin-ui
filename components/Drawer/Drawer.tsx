import * as React from 'react';

// Models
import { IBoxComponentProps } from '../Box/model';
import { IDrawerComponentProps, IDrawerContentComponentProps, IDrawerOverlayComponentProps } from './model';

// Styled Components
import {
    StyledDrawer,
    StyledDrawerBody,
    StyledDrawerCloseButton,
    StyledDrawerContentBottomIn,
    StyledDrawerContentBottomOut,
    StyledDrawerContentLeftIn,
    StyledDrawerContentLeftOut,
    StyledDrawerContentRightIn,
    StyledDrawerContentRightOut,
    StyledDrawerContentTopIn,
    StyledDrawerContentTopOut,
    StyledDrawerFooter,
    StyledDrawerHeader,
    StyledDrawerOverlayFadeIn,
    StyledDrawerOverlayFadeOut,
} from './styles';

const ContentByPosition = {
    bottom: {
        in: StyledDrawerContentBottomIn,
        out: StyledDrawerContentBottomOut
    },
    left: {
        in: StyledDrawerContentLeftIn,
        out: StyledDrawerContentLeftOut
    },
    right: {
        in: StyledDrawerContentRightIn,
        out: StyledDrawerContentRightOut
    },
    top: {
        in: StyledDrawerContentTopIn,
        out: StyledDrawerContentTopOut
    }
};

export const Drawer: React.FC<IDrawerComponentProps> = (props) => {

    const batchClickHandler = (e: React.MouseEvent<HTMLElement>, callback?: (e: React.MouseEvent<HTMLElement>) => void) => {
        props.onClose && props.onClose(e);
        callback && callback(e);
    };

    const recurseChildren = (children: any): any => {
        if (!children){
            return null;
        }

        return React.Children.map(children, (child, i) => {
            if (React.isValidElement<IBoxComponentProps>(child) && child.type === DrawerCloseButton){
                return React.cloneElement(child, {
                    children: recurseChildren(child.props.children),
                    key: i,
                    onClick: (e) => batchClickHandler(e, child.props.onClick)
                } as IBoxComponentProps);
            }

            if (React.isValidElement<IDrawerOverlayComponentProps>(child) && child.type === DrawerOverlay){
                return React.cloneElement(child, {
                    children: recurseChildren(child.props.children),
                    isOpen: child.props.isOpen ? child.props.isOpen : props.isOpen,
                    key: i,
                    onClick: props.closeOnOverlayClick ?
                                (e) => batchClickHandler(e, child.props.onClick) :
                                child.props.onClick
                } as IDrawerOverlayComponentProps);
            }

            if (React.isValidElement<IDrawerContentComponentProps>(child) && child.type === DrawerContent){
                return React.cloneElement(child, {
                    children: recurseChildren(child.props.children),
                    isOpen: child.props.isOpen ? child.props.isOpen : props.isOpen,
                    key: i,
                    onClick: (e: React.MouseEvent<HTMLElement>) => {
                        e.stopPropagation();
                        child.props.onClick && child.props.onClick(e);
                    },
                    placement: child.props.placement ? child.props.placement : props.placement,
                    size: child.props.size ? child.props.size : props.size,
                } as IDrawerContentComponentProps);
            }

            if (React.isValidElement(child) && (child.props as any).children){
                return React.cloneElement(child, {key: i, children: recurseChildren((child.props as any).children)} as any);
            }

            return child;
        });
    };

    return <StyledDrawer {...props} children={recurseChildren(props.children)} />;
};

Drawer.defaultProps = {
    className: '',
    closeOnOverlayClick: true,
    placement: 'right',
    size: 'sm',
    style: {}
};

Drawer.displayName = 'Drawer';

/**
 * DrawerContent
 */
export const DrawerContent: React.FC<IDrawerContentComponentProps> = (props) => {
    const DrawerByPlacement = props.placement && ContentByPosition[props.placement];

    if (!DrawerByPlacement){
        return null;
    }

    const DrawerContentComponent = props.isOpen ? DrawerByPlacement.in : DrawerByPlacement.out;

    return <DrawerContentComponent {...props} />;
};

DrawerContent.defaultProps = {
    className: '',
    style: {}
};

DrawerContent.displayName = 'DrawerContent';

/**
 * DrawerHeader
 */
export const DrawerHeader: React.FC<IBoxComponentProps> = (props) => {
    return <StyledDrawerHeader boxAs="header" {...props} />;
};

DrawerHeader.defaultProps = {
    className: '',
    style: {}
};

DrawerHeader.displayName = 'DrawerHeader';

/**
 * DrawerBody
 */
export const DrawerBody: React.FC<IBoxComponentProps> = (props) => {
    return <StyledDrawerBody {...props} />;
};

DrawerBody.defaultProps = {
    className: '',
    style: {}
};

DrawerBody.displayName = 'DrawerBody';

/**
 * DrawerFooter
 */
export const DrawerFooter: React.FC<IBoxComponentProps> = (props) => {
    return <StyledDrawerFooter boxAs="footer" {...props} />;
};

DrawerFooter .defaultProps = {
    className: '',
    style: {}
};

DrawerFooter.displayName = 'DrawerFooter';

/**
 * DrawerCloseButton
 */
export const DrawerCloseButton: React.FC<IBoxComponentProps> = (props) => {
    return <StyledDrawerCloseButton {...props} children="close" />;
};

DrawerCloseButton .defaultProps = {
    className: '',
    style: {}
};

DrawerCloseButton.displayName = 'DrawerCloseButton';

/**
 * DrawerOverlay
 */
export const DrawerOverlay: React.FC<IDrawerOverlayComponentProps> = (props) => {

    const [show, setShow] = React.useState(props.isOpen);

    React.useEffect(() => {
        if (props.isOpen){
            props.onOpen && props.onOpen();
            setShow(true);
        } else {
            setTimeout(() => {
                setShow(false);
            }, 200);
        }
    }, [props.isOpen]);

    const DrawerOverlayComponent = props.isOpen ? StyledDrawerOverlayFadeIn : StyledDrawerOverlayFadeOut;

    if (show){
        return <DrawerOverlayComponent {...props} />;
    }

    return null;

};

DrawerOverlay.defaultProps = {
    className: '',
    style: {}
};

DrawerOverlay.displayName = 'DrawerOverlay';

export default Drawer;
