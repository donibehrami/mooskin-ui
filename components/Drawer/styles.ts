import styled, {keyframes} from 'styled-components';

// Models
import { IBoxComponentProps } from '../Box/model';
import {
    IContentSize,
    IDrawerComponentProps,
    IDrawerContentComponentProps,
    IDrawerOverlayComponentProps
} from './model';

// Components
import Box from '../Box/Box';

// import variables from '../_utils/globals/variables';

const overlayFadeIn = keyframes`
    0% {
        background-color: rgba(0, 0, 0, 0);
    }
    100% {
        background-color: rgba(0, 0, 0, 0.48);
    }
`;

const contentRightFadeIn = keyframes`
    0% {
        opacity: 0;
        transform: translateX(100%) translateZ(0px);
    }
    100% {
        opacity: 1;
        transform: translateX(0%) translateZ(0px);
    }
`;

const contentLeftFadeIn = keyframes`
    0% {
        opacity: 0;
        transform: translateX(-100%) translateZ(0px);
    }
    100% {
        opacity: 1;
        transform: translateX(0%) translateZ(0px);
    }
`;

const contentBottomFadeIn = keyframes`
    0% {
        opacity: 0;
        transform: translateY(100%) translateZ(0px);
    }
    100% {
        opacity: 1;
        transform: translateY(0%) translateZ(0px);
    }
`;

const contentTopFadeIn = keyframes`
    0% {
        opacity: 0;
        transform: translateY(-100%) translateZ(0px);
    }
    100% {
        opacity: 1;
        transform: translateY(0%) translateZ(0px);
    }
`;

export const getContentSize = (size?: IContentSize) => {
    switch (size) {
        case 'xs':
            return '20%';

        case 'sm':
            return '35%';

        case 'md':
            return '50%';

        case 'lg':
            return '65%';

        case 'xl':
            return '80%';

        case 'full':
            return '100%';

        default:
            break;
    }
};

export const StyledDrawer = styled(Box)<IDrawerComponentProps>``;

export const StyledDrawerContent = styled(Box)<IDrawerContentComponentProps>`
    position: fixed;
    width: 100%;
    max-width: 100%;
    max-height: 100vh;
    height: 100vh;
    display: flex;
    flex-direction: column;
    z-index: 1400;
    color: inherit;
    outline: 0px;
    background: rgb(255, 255, 255);
    opacity: 1;
    box-shadow: rgba(0, 0, 0, 0.1) 0px 10px 15px -3px, rgba(0, 0, 0, 0.05) 0px 4px 6px -2px;
`;

export const StyledDrawerContentRight = styled(StyledDrawerContent)`
    right: 0px;
    top: 0px;
    max-width: ${(props) => getContentSize(props.size)};
`;

export const StyledDrawerContentRightIn = styled(StyledDrawerContentRight)`
    animation: ${contentRightFadeIn} 0.3s;
`;

export const StyledDrawerContentRightOut = styled(StyledDrawerContentRight)`
    animation: ${contentRightFadeIn} 0.3s reverse;
`;

export const StyledDrawerContentLeft = styled(StyledDrawerContent)`
    left: 0px;
    top: 0px;
    max-width: ${(props) => getContentSize(props.size)};
`;

export const StyledDrawerContentLeftIn = styled(StyledDrawerContentLeft)`
    animation: ${contentLeftFadeIn} 0.3s;
`;

export const StyledDrawerContentLeftOut = styled(StyledDrawerContentLeft)`
    animation: ${contentLeftFadeIn} 0.3s reverse;
`;

export const StyledDrawerContentTop = styled(StyledDrawerContent)`
    left: 0px;
    top: 0px;
    max-height: ${(props) => getContentSize(props.size)};
`;

export const StyledDrawerContentTopIn = styled(StyledDrawerContentTop)`
    animation: ${contentTopFadeIn} 0.3s;
`;

export const StyledDrawerContentTopOut = styled(StyledDrawerContentTop)`
    animation: ${contentTopFadeIn} 0.3s reverse;
`;

export const StyledDrawerContentBottom = styled(StyledDrawerContent)`
    left: 0px;
    bottom: 0px;
    max-height: ${(props) => getContentSize(props.size)};
`;

export const StyledDrawerContentBottomIn = styled(StyledDrawerContentBottom)`
    animation: ${contentBottomFadeIn} 0.3s;
`;

export const StyledDrawerContentBottomOut = styled(StyledDrawerContentBottom)`
    animation: ${contentBottomFadeIn} 0.3s reverse;
`;

export const StyledDrawerHeader = styled(Box)<IBoxComponentProps>`
    flex: 0 1 0%;
    padding: 1rem 1.5rem;
    font-size: 1.25rem;
    font-weight: 600;
    width: 100%;
`;

export const StyledDrawerBody = styled(Box)<IBoxComponentProps>`
    padding: 0.5rem 1.5rem;
    flex: 1 1 0%;
    overflow: auto;
    width: 100%;
`;

export const StyledDrawerFooter = styled(Box)<IBoxComponentProps>`
    display: flex;
    align-items: center;
    justify-content: flex-end;
    padding: 1rem 1.5rem;
    width: 100%;
`;

export const StyledDrawerCloseButton = styled(Box)<IBoxComponentProps>`
    font-family: 'Mooskin Icons';
    width: 32px;
    height: 32px;
    font-size: 32px;
    cursor: pointer;
    border-radius: 9px;
    :hover {
        background: rgba(30, 30, 30, 0.1);
    }
`;

export const StyledDrawerOverlay = styled(Box)<IDrawerOverlayComponentProps>`
    display: flex;
    width: 100vw;
    height: 100vh;
    position: fixed;
    left: 0px;
    top: 0px;
    z-index: 1400;
    justify-content: center;
`;

export const StyledDrawerOverlayFadeIn = styled(StyledDrawerOverlay)`
    background-color: rgba(0, 0, 0, 0.48);
    animation: ${overlayFadeIn} 0.3s;
`;

export const StyledDrawerOverlayFadeOut = styled(StyledDrawerOverlay)`
    background-color: rgba(0, 0, 0, 0);
    animation: ${overlayFadeIn} 0.3s reverse;
`;
