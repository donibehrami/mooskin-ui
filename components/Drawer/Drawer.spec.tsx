import * as React from 'react';
import Drawer, {DrawerBody, DrawerCloseButton, DrawerContent, DrawerFooter, DrawerHeader, DrawerOverlay} from './Drawer';

import { mount } from 'enzyme';

describe('Drawer', () => {

    test('renders correctly', () => {
        const fn = jest.fn();
        const fn2 = jest.fn();

        const tree = mount(
            <Drawer onClose={fn}>
                <DrawerOverlay onClick={fn2}>
                    <DrawerContent w="50%" h="50%">
                        <DrawerCloseButton position="absolute" top={10} right={10} />
                        <DrawerHeader>Create your account</DrawerHeader>

                        <DrawerBody>
                            Drawer Content Body
                        </DrawerBody>

                        <DrawerFooter>
                            Drawer Footer goes here!
                        </DrawerFooter>
                    </DrawerContent>
                </DrawerOverlay>
            </Drawer>
        );

        expect(tree).toMatchSnapshot();
    });

});
