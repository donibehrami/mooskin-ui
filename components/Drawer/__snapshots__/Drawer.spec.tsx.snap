// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Drawer renders correctly 1`] = `
<Drawer
  className=""
  closeOnOverlayClick={true}
  onClose={[MockFunction]}
  placement="right"
  size="sm"
  style={Object {}}
>
  <Styled(Box)
    className=""
    closeOnOverlayClick={true}
    onClose={[MockFunction]}
    placement="right"
    size="sm"
    style={Object {}}
  >
    <Box
      className="sc-gtssRu"
      closeOnOverlayClick={true}
      onClose={[MockFunction]}
      placement="right"
      size="sm"
      style={Object {}}
    >
      <Component
        className="sc-gtssRu"
        closeOnOverlayClick={true}
        onClose={[MockFunction]}
        placement="right"
        size="sm"
        style={Object {}}
        theme={
          Object {
            "disabledBackground": "#e2e2e2",
            "disabledfont": "#9d9d9d",
            "inputBackground": "#ffffff",
            "inputBorder": "#e2e2e2",
            "inputFont": "#9d9d9d",
            "label": "#2d2d2d",
            "primary": "#3fbaca",
            "secondary": "#293346",
          }
        }
      >
        <styled.div
          className="sc-gtssRu"
          closeOnOverlayClick={true}
          onClose={[MockFunction]}
          placement="right"
          size="sm"
          style={Object {}}
          theme={
            Object {
              "disabledBackground": "#e2e2e2",
              "disabledfont": "#9d9d9d",
              "inputBackground": "#ffffff",
              "inputBorder": "#e2e2e2",
              "inputFont": "#9d9d9d",
              "label": "#2d2d2d",
              "primary": "#3fbaca",
              "secondary": "#293346",
            }
          }
        >
          <div
            className="sc-bdnylx dOPKTt sc-gtssRu"
            onClose={[MockFunction]}
            size="sm"
            style={Object {}}
          >
            <DrawerOverlay
              className=""
              key="0/.0"
              onClick={[Function]}
              style={Object {}}
            />
          </div>
        </styled.div>
      </Component>
    </Box>
  </Styled(Box)>
</Drawer>
`;
